core/jvmMain/src/kotlinx/serialization/internal/Caching.kt

```diff
@@ -40,38 +41,87 @@ internal actual fun <T> createParametrizedCache(factory: (KClass<Any>, List<KTyp
+ * In the case of serializers, these should be instances of the same class filled with equivalent values.
+ */
 @SuppressAnimalSniffer
-private class ClassValueWrapper<T>(private val compute: (KClass<*>) -> KSerializer<T>?): ClassValue<CacheEntry<T>>() {
-    /*
-     * Since during the computing of the value for the `ClassValue` entry, we do not know whether a nullable
-     *  serializer is needed, so we may need to differentiate nullable/non-null  caches by a level higher
-     */
-    override fun computeValue(type: Class<*>): CacheEntry<T> {
-        return CacheEntry(compute(type.kotlin))
+private class ClassValueReferences<T> : ClassValue<MutableSoftReference<T>>() {
+    override fun computeValue(type: Class<*>): MutableSoftReference<T> {
+        return MutableSoftReference()
     }
-}
 
-private class ClassValueParametrizedCache<T>(private val compute: (KClass<Any>, List<KType>) -> KSerializer<T>?) : ParametrizedSerializerCache<T> {
-    private val classValue = ParametrizedClassValueWrapper<T>()
+    inline fun getOrSet(key: Class<*>, crossinline factory: () -> T): T {
+        val ref: MutableSoftReference<T> = get(key)
+
+        /*
+          `reference` is not volatile field, old value with null ref may be read.
+         It's ok, actual value will be read in synchronized block.
+
+         It is unlikely that the `reference` with the old non-null value will be read - but even in this case
+        it should not be a big problem, because it does not lead to the creation of a new serializer and reuses an existing one
+         */
+        ref.reference.get()?.let { return it }
+
+        // go to the slow path and create serializer with blocking, also wrap factory block
+        return ref.getOrSetAtomic { factory() }
+    }
 
-    override fun get(key: KClass<Any>, types: List<KType>): Result<KSerializer<T>?> =
-        classValue[key.java].computeIfAbsent(types) { compute(key, types) }
 }
 
-@SuppressAnimalSniffer
-private class ParametrizedClassValueWrapper<T> : ClassValue<ParametrizedCacheEntry<T>>() {
+/**
+ * Wrapper over `SoftReference`, used  to store a mutable value.
+ *
+ * **[reference] is not thread-safe, external concurrency management is needed**
+ */
+private class MutableSoftReference<T> {
+    @JvmField
+    var reference: SoftReference<T> = SoftReference(null)
+
     /*
-    * Since during the computing of the value for the `ClassValue` entry, we do not know whether a nullable
-    *  serializer is needed, so we may need to differentiate nullable/non-null  caches by a level higher
-    */
-    override fun computeValue(type: Class<*>): ParametrizedCacheEntry<T> {
-        return ParametrizedCacheEntry()
+    It is important that the monitor for synchronized is the `MutableSoftReference` of a specific class
+    This way access to reference is blocked only for one serializable class, and not for all
+     */
+    @Synchronized
+    fun getOrSetAtomic(factory: () -> T): T {
```

Comment:
I'd avoid the word `Atomic` here as it has its own semantics, which implies CAS and other things, but not locks. `getOrSetSynchronized` or `getOrSetWithLock` should be just fine.